### ISSUE TITLE
2nd Level cache partially ignored if self referencing many-to-one used with query cache 

### ISSUE THREAD

**Issue by marius-klimantavicius** [nhibernate/nhibernate-core#3179]
We have a self-referencing mapping similar to:
```xml
  <class name="DocumentTemplate" table="DocumentTemplate">
    <cache usage="read-write" region="LongTerm"/>
    <id name="Id" column="DocumentTemplate" type="int">
      <generator class="identity" />
    </id>
  ...
    <many-to-one name="Header" fetch="select"/>
    <many-to-one name="Footer" fetch="select"/>
  ...
  <query name="GetAllDocumentTemplates">
    <![CDATA[
          select dt
          from DocumentTemplate dt
            left outer join fetch dt.Component dtC
          where (dt.IsDeleted = false)
          order by dt.Name
        ]]>
  </query>
  ...
  </class>
```

where `Header` and `Footer` are properties of itself (`public DocumentTemplate Header { get; set; }`).

We call the query in this manner:
```csharp
            return Session.GetNamedQuery("GetAllDocumentTemplates")
                .SetCacheable(true)
                .SetCacheRegion(CacheRegions.LongTerm)
                .SetFlushMode(FlushMode.Manual)
                .List<DocumentTemplate>();
```

Both query cache and 2nd level cache is enabled:
```xml
    <property name="cache.use_second_level_cache">true</property>
    <property name="cache.use_query_cache">true</property>
```

The issue is that this query is only partially hitting 2nd level cache if entities are loaded in similar order:
```json
[
  { "Id": 1, "Header": 2, "Footer": null },
  { "Id": 2, "Header": null, "Footer": null }
]
```

The entity with `Id: 1` is loaded from 2nd level cache, however during the load a database request for entity `Id: 2` is issued even though I'd expect it to be also cached in 2nd level cache (from previous call).

As far as I was able to debug the following happens:
1. `BatchFetchQueue` is used to load cached entities
2. The entities are marked (`_queryCheckedEntityKeys`) as being loaded
3. During batch load (https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/Event/Default/DefaultLoadEventListener.cs#L460) an entity is being assembled which does not have `Header` or `Footer` loaded yet (it is returned by the query as all rows are loaded - these are not marked as `IsDeleted`)
4. `OneToManyType` issues a load for the property but the referenced entity it is marked as "checked" (and is supposed to be loaded later by batch loader) thus 2nd level cache is skipped and a request to database issued

**Comment by bahusoid** [comment]
Is it reproducible with latest 5.3.13? You need to provide a reproducible test case (you can use https://github.com/nhibernate/nhibernate-core-testcase/ see test template https://github.com/nhibernate/nhibernate-core-testcase/tree/master/NHSpecificTest/GH0000)

**Comment by marius-klimantavicius** [comment]
While trying to make a test case I found that the issue with our code was that our class was not marked as `lazy`. I've omitted `default-lazy=true` in the initial description as I did not think it was relevant.

Once I marked my class as lazy (or removed `default-lazy="false"`) then the additional requests to database disappeared and thus the entities were fully loaded from 2nd level cache.

Closing as works as designed.
